### ISSUE TITLE
AIC

### ISSUE THREAD

**Issue by yaxigeigei** [openconnectome/Statistical-Connectomics-Class#204]
In my final project I used the SBM and the number of cluster, k, is a parameter I need to optimize. 
Let's take Jovo's demo MATLAB code as an example:

// Sample an Adjacency Matrix from SBM(k = 1)
p = 0.1;                     // connection prob.
n = 100;                    // number of vertices
A = rand(n) < p;        // sampling

// Log-likelihood
loglikA = A*log(p) + (1-A)*log(1-p);
sumloglikA = sum(loglikA(:));

If I want to add a penalty term, for instance using AIC, what would that look like?
// Like this? 
d = 1;   // number of parameter
aic = -2 * sumloglikeA + 2*d;

**Comment by whock** [comment]
Yes, I think that's the right way of doing AIC. While double-checking the formula online I also found something called AICc (AIC corrected) which is a modified form of the equation that takes into account finite sample sizes. 

Here it is:
![image](https://cloud.githubusercontent.com/assets/10751093/7104108/50fbadd4-e097-11e4-8414-1aae19246609.png)

The reason to use AICc is that when n is on the same order as k^2 (i.e. you have comparable number of parameters and data points) then the risk of overfitting becomes very large, even with the built-in penalty term of AIC. So with few data points or a complex model, AICc should be used. (And if n is big AICc reduces back to regular AIC so it's usually the best option regardless)

Sources: 

Burnham, K. P.; Anderson, D. R. (2002), Model Selection and Multimodel Inference: A Practical Information-Theoretic Approach (2nd ed.), Springer-Verlag, ISBN 0-387-95364-7.

Burnham, K. P.; Anderson, D. R. (2004), "Multimodel inference: understanding AIC and BIC in Model Selection", Sociological Methods & Research 33: 261–304, doi:10.1177/0049124104268644

Claeskens, G.; Hjort, N. L. (2008), Model Selection and Model Averaging, Cambridge University Press.

**Comment by yaxigeigei** [comment]
Thank you for looking this up. So, in the demo example, what is the n?

**Comment by whock** [comment]
Would it be 100? Maybe a quant person can weigh in.

**Comment by mrjiaruiwang** [comment]
you could try a cross validation and plot the errors for each value of k you test and then find the one that has minimal error?

**Comment by yaxigeigei** [comment]
Well, the cross validation sounds good although I have no concrete idea about the substrate of validation. Would you please write some simple pseudo code based on the toy example? Also, I do want to know how to properly calculate AIC in this situation.

**Comment by whock** [comment]
@yaxigeigei I also would like to know how to calculate AIC correctly - I've learned it before as just the normal AIC equation but now I'm not sure which is more appropriate, AIC or AICc

And for the cross-validation issue, what exactly do you mean by substrate? If you're referring what to perform the CV on I don't think splitting the nodes into n groups for validation is a good idea because then you'd have fundamentally different graphs in each set. Maybe multiple draws from an SBM distribution? But then the dataset partitions wouldn't actually be the same (because they come from different samplings)? I agree I don't know exactly what the substrate would be.

**Comment by yaxigeigei** [comment]
Great questions eliciting more questions!
